src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -0,0 +1,345 @@
+  auto node = gcs_actor_scheduling_strategy_->SelectNode(actor);
+  if (node == nullptr) {
+    RAY_LOG(WARNING) << "Failed to select node to create actor " << actor->GetActorID()
+                     << " enqueue the actor and recreate it later.";
+    Enqueue(actor);
+    return;
+  }
+
+  // Update the node id and flush the actor table data to storage asynchronously.
+  actor->SetNodeID(node->GetNodeId());
+  std::weak_ptr<GcsNode> weak_node(node);
+  AsyncFlush(actor, [this, actor, weak_node] {
+    if (auto node = weak_node.lock()) {
+      // Lease worker from node if the node is still available.
+      LeaseWorker(actor, node);
+    } else {
+      RAY_LOG(WARNING) << "Failed to create actor " << actor->GetActorID() << " as node "
+                       << actor->GetNodeID()
+                       << " is dead, enqueue the actor and recreate it later.";
+      actor->SetNodeID(ClientID::Nil());
+      // Since the actor is not yet bind with leased worker, so we should enqueue the
+      // actor otherwise it will never be created later.
+      Enqueue(actor);
+    }
+  });
+}
+
+void GcsActorManager::LeaseWorker(std::shared_ptr<GcsActor> actor,
+                                  std::shared_ptr<GcsNode> node) {
+  RAY_CHECK(actor && node);
+  RAY_LOG(DEBUG) << "Start leasing worker from node " << node->GetNodeId()
+                 << " for actor " << actor->GetActorID();
+
+  std::weak_ptr<GcsNode> weak_node(node);
+  node->LeaseWorker(actor, [this, actor, weak_node](
+                               const Status &status, const ClientID &spillback_to,
+                               std::shared_ptr<GcsLeasedWorker> worker) {
+    if (auto node = weak_node.lock()) {
+      if (!status.ok()) {
+        if (auto gcs_node_manager = gcs_node_manager_.lock()) {
+          // TODO(zsl): Treat it as NodeFailure when !status.ok(), it may be a little
+          // strict. Maybe better to retry leasing worker from this node.
+          gcs_node_manager->OnNodeFailure(node->GetNodeId());
+        }
+        return;
+      }
+
+      if (!spillback_to.IsNil()) {
+        RAY_CHECK(worker == nullptr);
+        actor->SetNodeID(spillback_to);
+        AsyncFlush(actor, [this, actor] { Enqueue(actor); });
+        return;
+      }
+
+      RAY_LOG(INFO) << "Succeed in leasing worker " << worker->GetWorkerID() << " from "
+                    << worker->GetNodeID() << " for actor " << actor->GetActorID();
+      OnWorkerLeased(actor, std::move(worker));
+    }
+  });
+}
+
+void GcsActorManager::OnWorkerLeased(std::shared_ptr<GcsActor> actor,
+                                     std::shared_ptr<GcsLeasedWorker> worker) {
+  RAY_CHECK(actor && worker && actor->GetWorker() == nullptr);
+  RAY_LOG(INFO) << "Start creating actor " << actor->GetActorID() << " on worker "
+                << worker->GetWorkerID() << " at node " << actor->GetNodeID();
+  std::weak_ptr<GcsLeasedWorker> weak_worker(worker);
```

Comment:
Why a weak ptr?

Reply:
Actually, the life time of `GcsLeasedWorker` is managed by `GcsNode`. The `GcsActorManager` just reference the weak_ptr of  each leased worker.

Consider that if the the `CreateActor` is executing asynchronously and waiting the reply of worker, at this time the GCS Server detect that the raylet is crashed, so all of the leased worker associated with the dead GcsNode will be destroyed. At this time, if the callback of  `CreateActor` is triggered, we could just ignore the callback as the weak_ptr locks nothing.

Reply:
- The purpose of using the weak_ptr here is to make the worker's life cycle easier to manage.
- Capturing the shared_ptr in lambda is another choice, but it is a little more complexity because it need add a flag to the worker, if the worker is dead we should mark the flag to dead, and in the callback of IO event we should do like this:
    ``` if (!shared_worker->is_dead) { do_something(); }```